### ISSUE TITLE
PowerPC 32 bit support

### ISSUE THREAD

**Issue by travitch** [GaloisInc/pate#13]
I believe the current PowerPC backend for the `pate` tool supports ppc64. I suspect we will need ppc32 for the embedded challenge problems. This ticket will track the sub-tasks required to make this possible:
- [ ] Update macaw-loader-ppc (the 32 bit version) to not look for the `.opd` section, as ppc32 does not use a Table of Contents
- [ ] Investigate the reliance of ppc32 on a Small Data Area (SDA) pointer, which is held in the reserved `r13`; this may not affect all ABIs. It is not obvious where this register gets populated.
- [ ] In handling the SDA pointer, we may need to generalize how pate handles dedicated registers (e.g., the TOC register for ppc64); currently it has very special handling for the TOC, but might be better served by a generic API that just provides block or function-level initializers for dedicated registers
- [ ] Separate the `handleSystemCall` and `handleExternalCall` `ExternalDomain`s for ppc32 and ppc64
- [ ] Add tests for ppc32

**Comment by travitch** [comment]
As a sketch for the generalization of TOC handling, we could:
- [ ] Remove `tocProof` from `ValidArch`
- [ ] Remove `HasTOCReg`
- [ ] Remove `HasTOCDict`
- [ ] Remove `withTOCCases`
- [ ] In `RegisterCases`, replace `RegTOC` with `RegDedicated (DedicatedRegister arch)`, where `DedicatedRegister` is a type family
- [ ] Provide a function that can be called from `validRegister` in Validity.hs that can compute the necessary validity condition for the dedicated register in the original and patched programs (generalizing the TOC code we currently have there)
- [ ] Evaluate the needs in `guessEquivalenceDomain`; we probably want dedicated registers to always be excluded, but we may want to make that configurable

**Comment by travitch** [comment]
It looks like there are multiple ABI options for 32 bit PowerPC. The default gcc options (at least on my machine...) do not do anything special and construct all global data addresses using a two instruction sequence (`li` followed by `lis`). Building with `-meabi` enables the use of `r2` and `r13` as special dedicated registers from which global offsets are computed. In that mode, `r13` is suitably initialized, but not actually used; it is only used if the `-msdata` flag is specified (to enable the small data section optimization).

Long story short, it isn't clear what ABI(s) we'll ultimately need to support. I'll close this ticket once we get ppc32 working *without* eabi support. If we need it, we can add it later.
